programs/bpf_loader/src/syscalls.rs

```diff
@@ -115,174 +115,254 @@ impl SyscallConsume for Rc<RefCell<ComputeMeter>> {
+        SyscallKeccak256::init,
+        SyscallKeccak256::call,
+    )?;
 
-    if invoke_context
-        .feature_set
-        .is_active(&blake3_syscall_enabled::id())
-    {
-        syscall_registry.register_syscall_by_name(b"sol_blake3", SyscallBlake3::call)?;
-    }
+    // Secp256k1 Recover
+    register_feature_gated_syscall!(
+        syscall_registry,
+        secp256k1_recover_syscall_enabled,
+        b"sol_secp256k1_recover",
+        SyscallSecp256k1Recover::init,
+        SyscallSecp256k1Recover::call,
+    )?;
 
-    if invoke_context
-        .feature_set
-        .is_active(&zk_token_sdk_enabled::id())
-    {
-        syscall_registry
-            .register_syscall_by_name(b"sol_zk_token_elgamal_op", SyscallZkTokenElgamalOp::call)?;
-        syscall_registry.register_syscall_by_name(
-            b"sol_zk_token_elgamal_op_with_lo_hi",
-            SyscallZkTokenElgamalOpWithLoHi::call,
-        )?;
-        syscall_registry.register_syscall_by_name(
-            b"sol_zk_token_elgamal_op_with_scalar",
-            SyscallZkTokenElgamalOpWithScalar::call,
-        )?;
-    }
+    // Blake3
+    register_feature_gated_syscall!(
+        syscall_registry,
+        blake3_syscall_enabled,
+        b"sol_blake3",
+        SyscallBlake3::init,
+        SyscallBlake3::call,
+    )?;
+
+    // ZK Token
+    register_feature_gated_syscall!(
+        syscall_registry,
+        zk_token_sdk_enabled,
+        b"sol_zk_token_elgamal_op",
+        SyscallZkTokenElgamalOp::init,
+        SyscallZkTokenElgamalOp::call,
+    )?;
+    register_feature_gated_syscall!(
+        syscall_registry,
+        zk_token_sdk_enabled,
+        b"sol_zk_token_elgamal_op_with_lo_hi",
+        SyscallZkTokenElgamalOpWithLoHi::init,
+        SyscallZkTokenElgamalOpWithLoHi::call,
+    )?;
+    register_feature_gated_syscall!(
+        syscall_registry,
+        zk_token_sdk_enabled,
+        b"sol_zk_token_elgamal_op_with_scalar",
+        SyscallZkTokenElgamalOpWithScalar::init,
+        SyscallZkTokenElgamalOpWithScalar::call,
+    )?;
 
-    syscall_registry
-        .register_syscall_by_name(b"sol_get_clock_sysvar", SyscallGetClockSysvar::call)?;
+    // Sysvars
+    syscall_registry.register_syscall_by_name(
+        b"sol_get_clock_sysvar",
+        SyscallGetClockSysvar::init,
+        SyscallGetClockSysvar::call,
+    )?;
     syscall_registry.register_syscall_by_name(
         b"sol_get_epoch_schedule_sysvar",
+        SyscallGetEpochScheduleSysvar::init,
         SyscallGetEpochScheduleSysvar::call,
     )?;
-    if !invoke_context
-        .feature_set
-        .is_active(&disable_fees_sysvar::id())
-    {
-        syscall_registry
-            .register_syscall_by_name(b"sol_get_fees_sysvar", SyscallGetFeesSysvar::call)?;
-    }
-    syscall_registry
-        .register_syscall_by_name(b"sol_get_rent_sysvar", SyscallGetRentSysvar::call)?;
+    register_feature_gated_syscall!(
+        syscall_registry,
+        disable_fees_sysvar,
```

Comment:
@Lichtso I suspect this is the issue; invert of the logic on the `disable_fees_sysvar`, should be `!disable_fees_sysvar`.

Reply:
Yes, looks promising. Can you test that?

- Checkout 6bbfef7069e7d7a05573fa0ddb25e60ec22b55eb the commit before #24694 made it in
- Revert the revert (#24647)
- Fix this boolean inversion
- Test by running a validator

Reply:
#24546